lib/graphql/dataloader/loader.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+module GraphQL
+  class Dataloader
+    class Loader
+      class StateError < GraphQL::Error; end
+
+      def self.load(context, key, value)
+        dl = context[:dataloader]
+        loader = dl.loaders[self][key]
+        loader.load(value)
+      end
+
+      def initialize(context, key)
+        @context = context
+        @key = key
+        @promises = {}
+        @loaded_values = {}
+      end
+
+      def load(value)
+        @promises[value] ||= GraphQL::Execution::Lazy.new do
+          if !@loaded_values.key?(value)
+            sync
```

Comment:
In the concurrent example, `sync` will run in a separate thread. Correct?

If so, then how do you guarantee that `@loaded_values[value]` will be present on line 26 below?